source/common/ssl/context_impl.cc

```diff
@@ -334,13 +354,41 @@ bool ContextImpl::dNSNameMatch(const std::string& dNSName, const char* pattern)
 }
 
 bool ContextImpl::verifyCertificateHashList(
-    X509* cert, const std::vector<std::vector<uint8_t>>& certificate_hash_list) {
+    X509* cert, const std::vector<std::vector<uint8_t>>& expected_hashes) {
   std::vector<uint8_t> computed_hash(SHA256_DIGEST_LENGTH);
   unsigned int n;
   X509_digest(cert, EVP_sha256(), computed_hash.data(), &n);
   RELEASE_ASSERT(n == computed_hash.size());
 
-  for (const auto& expected_hash : certificate_hash_list) {
+  for (const auto& expected_hash : expected_hashes) {
+    if (computed_hash == expected_hash) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool ContextImpl::verifyCertificateSpkiList(
+    X509* cert, const std::vector<std::vector<uint8_t>>& expected_hashes) {
+  X509_PUBKEY* pubkey = X509_get_X509_PUBKEY(cert);
+  if (pubkey == nullptr) {
+    return false;
+  }
+  // 1st pass to get the length, 2nd pass to write DER-encoded SPKI to the stack-allocated "spki".
+  const int len = i2d_X509_PUBKEY(pubkey, nullptr);
+  if (len < 0) {
+    return false;
+  }
+  uint8_t spki[len];
+  uint8_t* spkip = spki;
+  if (len != i2d_X509_PUBKEY(pubkey, &spkip)) {
```

Comment:
Should this be a RELEASE_ASSERT for this pass? Can this ever fail? Can the one above ever really fail (I don't know the details on this)?

Reply:
I don't think `i2d_TYPE()` can fail after it successfully returned length in the first pass, and there shouldn't be any extra allocations, but I'm not 100% sure. @davidben do you know?

Reply:
I'm not positive. The legacy ASN.1 and X.509 code is in one of those "deprecated but replacement isn't fully ready" states, in part because that code is a mess that no one understands. I suspect `malloc` failures are possible, but I expect ones involving invalid input are not.

But this is less typing:

```
  uint8_t *spki = nullptr;
  const int len = i2d_X509_PUBKEY(pubkey, &spki);
  if (len < 0) {
    return false;
  }
  bssl::UniquePtr<uint8_t> free_spki(spki);  // This just calls OPENSSL_free.

  std::vector<uint8_t> computed_hash(...);
  ...
```

The `i2d` functions are absurd. (There's a reason we're working to abandon this part of the library... it's pretty much unsalvageable.) If the pointer starts out as `NULL`, it goes through a totally different calling convention where it allocates the buffer for you. That'll also avoid the variable-length arrays, which is a GCC-ism I think, as well as avoid potentially allowing an attacker to make you allocate more data on the stack than you have room for.

Reply:
Yeah, I originally had this version (albeit with direct call to `OPENSSL_free`), but I switched to 2-step version during review to avoid dealing with memory allocated by BoringSSL. I didn't realize that VLAs never made it to C++ (they are part of C99), so I reverted back to your version. Thanks!